Vol2 hpo (#2372)
d2l-ai/d2l-en: chapter_hyperparameter_optimization/sh-async.md

![Visualization of successive halving with synchronous and asynchronous scheduling of trials.](img/asha.svg)
:label:`asha`

As for asynchronous random search, we will use *Syne Tune* once more.

## Objective Function

Just as in :numref:`sec_rs_async`, we will use Syne Tune with the following
objective function.

```{.python .input  n=54}
import logging
logging.basicConfig(level=logging.INFO)
import matplotlib.pyplot as plt

from syne_tune import Tuner, StoppingCriterion
from syne_tune.backend.python_backend import PythonBackend
from syne_tune.config_space import randint, loguniform
from syne_tune.optimizer.baselines import ASHA
from syne_tune.experiments import load_experiment

def objective(learning_rate, batch_size, max_epochs):
    from d2l import torch as d2l    
    from syne_tune import Reporter
    model = d2l.AlexNet(lr=learning_rate)
    trainer = d2l.Trainer(max_epochs=1, num_gpus=1)
    data = d2l.FashionMNIST(batch_size=batch_size, resize=(224, 224))
    report = Reporter() 
    for epoch in range(1, max_epochs + 1):
        trainer.fit(model=model, data=data)
        validation_error = d2l.numpy(trainer.validate().cpu())

>         report(epoch=epoch-1, validation_error=float(validation_error))

Reviewer: ```suggestion
    for epoch in range(max_epochs):
        trainer.fit(model=model, data=data)
        validation_error = d2l.numpy(trainer.validate().cpu())
        report(epoch=epoch, validation_error=float(validation_error))
```
We can probably simplify this, right?

```

We also use the same configuration space as before

```{.python .input  n=55}
max_epochs = 4

config_space = {
   "learning_rate": loguniform(1e-5, 1e-1),
   "batch_size": randint(8, 128),
   "max_epochs": max_epochs,
}
```

```{.python .input}
# We have to set this number equal to the number of GPUs that are in the machine
# to run this notebook
n_workers = 2
max_wallclock_time = 15 * 60
```

## Asynchronous Scheduler

The code for running ASHA is a simple variation of what we did for asynchronous
random search.

```{.python .input  n=56}
scheduler = ASHA(
    config_space,
    metric="validation_error",
    mode="min",
    max_resource_attr="max_epochs",
    resource_attr="epoch",
    type="promotion",
    grace_period=1,  # this corresponds to r_min 

>     reduction_factor=2,  # this corresponds to eta

Reviewer: ```suggestion
    reduction_factor=2  # this corresponds to eta
```

)  
```

Here, `metric` and `resource_attr` specify the key names used with the `report`
callback, and `max_resource_attr` denotes which input to the objective function
corresponds to $r_{max}$. Moreover, `grace_period` provides $r_{min}$, and
`reduction_factor` is $\eta$.

Now, we can run Syne Tune as before:

```{.python .input  n=57}
trial_backend = PythonBackend(tune_function=objective, config_space=config_space)

stop_criterion = StoppingCriterion(
    max_wallclock_time=max_wallclock_time
)
tuner = Tuner(
    trial_backend=trial_backend,
    scheduler=scheduler,
    stop_criterion=stop_criterion,

>     n_workers=n_workers,

Reviewer: ```suggestion
    n_workers=n_workers
```
